File: app/src/main/java/com/jml/rappichallenge/models/entities/Movie.java
package com.jml.rappichallenge.models.entities;

import android.content.Context;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.jml.rappichallenge.models.tools.LanguageHelper;

import java.lang.reflect.Type;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class Movie
{
	private static class DTO
	{
		private int vote_count;
		private int id;
		private boolean video;
		private float vote_average;
		private String title;
		private float popularity;
		private String poster_path;
		private String backdrop_path;
		private String original_language;
		private String original_title;
		private List<Namable> genres;
		private boolean adult;
		private String overview;
		private String release_date;
		private int budget;
		private int revenue;
		private String homepage;
		private String imdbId;
		private int runtime;
		private List<Namable> production_companies; // TODO <- DTO mas completo tiene imagen
		private List<Namable> production_countries;
		private List<Namable> spoken_languages;
		private String status;
		private String tagline;
	}

	private DTO dto;

	Movie(DTO dto)
	{
		this.dto = dto;
	}

	public int getId()
	{
		return dto.id;
	}

	public boolean hasVideo()
	{
		return dto.video;
	}

	public float getVoteCount()
	{
		return dto.vote_count;
	}

	public float getVoteAverage()
	{
		return dto.vote_average;
	}

	public String getTitle()
	{
		return dto.title;
	}

	public float getPopularity()
	{
		return dto.popularity;
	}

	public String getPosterPath()
	{
		return dto.poster_path;
	}

	public String getBackdropPath()
	{
		return dto.backdrop_path;
	}

	public String getOriginalLanguageCode()
	{
		return dto.original_language;
	}

	public String getOriginalTitle()
	{
		return dto.original_title;
	}


	// TODO Generos
	//private List<Integer> genre_ids;

	public boolean isAdult()
	{
		return dto.adult;
	}

	public String getOverview()
	{
		return dto.overview;
	}

	public Date getReleaseDate()
	{
		DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());

		try
		{
			return dateFormat.parse(dto.release_date);
		}
		catch (ParseException e)
		{
			return null;
		}
	}

	public List<Namable> getGenres()
	{
		return new ArrayList<>(dto.genres);
	}

	public int getBudget()
	{
		return dto.budget;
	}

	public int getRevenue()
	{
		return dto.revenue;
	}

	public String getHomepage()
	{
		return dto.homepage;
	}

	public String getImdbId()
	{
		return dto.imdbId;
	}

	public int getRuntime()
	{
		return dto.runtime;
	}

	public List<Namable> getProductionCompanies()
	{
		return new ArrayList<>(dto.production_companies);
	}

	public List<Namable> getProductionCountries()
	{
		return new ArrayList<>(dto.production_countries);
	}

	public List<Namable> getSpokenLanguages()
	{
		return new ArrayList<>(dto.spoken_languages);
	}

	public String getStatus()
	{
		return dto.status;
	}

	public String getTagline()
	{
		return dto.tagline;
	}

	public static class Deserializer implements JsonDeserializer<Movie>
	{
		@Override
		public Movie deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
		{
			Movie.DTO dto = context.deserialize(json, Movie.DTO.class);
			return new Movie(dto);
		}
	}

	public static class FormatHelper
	{
		public static String getFormmatedLanguage(Context context, String languageCode)
		{
			int id = LanguageHelper.GetLanguageResNameByCode(languageCode);

			if(id == 0)
			{
				return "";
			}

			return context.getString(id);
		}
	}
}

File: app/src/main/java/com/jml/rappichallenge/models/entities/Movie.kt
package com.jml.rappichallenge.models.entities

import android.content.Context

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.jml.rappichallenge.models.dtos.MovieDTO
import com.jml.rappichallenge.models.tools.LanguageHelper

import java.lang.reflect.Type
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

class Movie internal constructor(internal val dto: MovieDTO) {

    val id: Int
        get() = dto.id

    val voteCount: Float
        get() = dto.vote_count.toFloat()

    val voteAverage: Float?
        get() = dto.vote_average

    val title: String?
        get() = dto.title

    val popularity: Float
        get() = dto.popularity

    val posterPath: String?
        get() = dto.poster_path

    val backdropPath: String?
        get() = dto.backdrop_path

    val originalLanguageCode: String?
        get() = dto.original_language

    val originalTitle: String?
        get() = dto.original_title


    // TODO Generos
    //private List<Integer> genre_ids;

    val isAdult: Boolean
        get() = dto.adult

    val overview: String?
        get() = dto.overview

    val releaseDate: Date?
        get() {
            val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

            try {
                return dateFormat.parse(dto.release_date)
            } catch (e: ParseException) {
                return null
            }

        }

    val genres: List<Namable>
        get() = dto.genres?.map { namableDTO -> Namable(namableDTO) } ?: ArrayList()

    val budget: Int
        get() = dto.budget

    val revenue: Int
        get() = dto.revenue

    val homepage: String?
        get() = dto.homepage

    val imdbId: String?
        get() = dto.imdbId

    val runtime: Int
        get() = dto.runtime

    val productionCompanies: List<Namable>
        get() = dto.production_companies?.map { namableDTO -> Namable(namableDTO) } ?: ArrayList()

    val productionCountries: List<Namable>
        get() = dto.production_countries?.map { namableDTO -> Namable(namableDTO) } ?: ArrayList()

    val spokenLanguages: List<Namable>
        get() = dto.spoken_languages?.map { namableDTO -> Namable(namableDTO) } ?: ArrayList()

    val status: String?
        get() = dto.status

    val tagline: String?
        get() = dto.tagline

    fun hasVideo(): Boolean {
        return dto.video
    }

    class Deserializer : JsonDeserializer<Movie> {
        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Movie {
            val dto = context.deserialize<MovieDTO>(json, MovieDTO::class.java)
            return Movie(dto)
        }
    }

    object FormatHelper {
        fun getFormmatedLanguage(context: Context, languageCode: String): String {
            val id = LanguageHelper.getLanguageResNameByCode(languageCode)

            return if (id == 0) {
                ""
            } else context.getString(id)

        }
    }
}
